server/gateway.go

```diff
@@ -0,0 +1,1707 @@
+	)
+	for _, c := range gws {
+		// Always send for queues. Note that by contract this function
+		// is called only once per account/subject/queue.
+		sendProto := sub.queue != nil
+		// We still want to execute following code even if this is a
+		// queue and we are going to send the proto anyway, because
+		// it may clear the noInterest map if we are going from
+		// a no-interest to a queue sub.
+		if sni, _ := c.gw.noInterest.Load(accName); sni != nil {
+			sn := sni.(*sync.Map)
+			if _, noInterest := sn.Load(string(sub.subject)); noInterest {
+				sn.Delete(string(sub.subject))
+				sendProto = true
+			}
+		}
+		if sendProto {
+			if len(proto) == 0 {
+				proto = append(proto, rSubBytes...)
+				proto = append(proto, accName...)
+				proto = append(proto, ' ')
+				proto = append(proto, sub.subject...)
+				if sub.queue != nil {
+					proto = append(proto, ' ')
+					proto = append(proto, sub.queue...)
+					// For now, just use 1 for the weight
+					proto = append(proto, ' ', '1')
+				}
+				proto = append(proto, CR_LF...)
+			}
+			c.mu.Lock()
+			if c.trace {
+				c.traceOutOp("", proto[:len(proto)-LEN_CR_LF])
+			}
+			c.sendProto(proto, true)
+			c.mu.Unlock()
+		}
+	}
+}
+
+func (s *Server) sendQueueUnsubToGateways(accName string, qsub *subscription) {
+	// If this is a remote queue sub, need to remove from remote queue subs map
+	if qsub.client != nil && qsub.client.typ == ROUTER {
+		s.gateway.Lock()
+		delete(s.gateway.rqs, string(qsub.sid))
+		s.gateway.Unlock()
+	}
+	var (
+		_gws [4]*client
+		gws  = _gws[:0]
+	)
+	s.getInboundGateways(&gws)
+	if len(gws) == 0 {
+		return
+	}
+	var (
+		protoa = [512]byte{}
+		proto  = protoa[:0]
+	)
+	proto = append(proto, rUnsubBytes...)
+	proto = append(proto, accName...)
+	proto = append(proto, ' ')
+	proto = append(proto, qsub.subject...)
+	proto = append(proto, ' ')
+	proto = append(proto, qsub.queue...)
+	proto = append(proto, CR_LF...)
+	for _, c := range gws {
+		c.mu.Lock()
+		if c.trace {
+			c.traceOutOp("", proto[:len(proto)-LEN_CR_LF])
+		}
+		c.sendProto(proto, true)
+		c.mu.Unlock()
+	}
+}
+
+// May send a message to all outbound gateways. It is possible
+// that message is not sent to a given gateway if for instance
+// it is known that this gateway has no interest in account or
+// subject, etc..
+func (c *client) sendMsgToGateways(msg, subject, reply []byte, doQueues bool) {
+	var (
+		gwsa [4]*client
+		gws  = gwsa[:0]
+		s    = c.srv
+	)
+	s.getOutboundGateways(&gws)
+	if len(gws) == 0 {
+		return
+	}
+	var (
+		subj    = string(subject)
+		queuesa = [512]byte{}
+		queues  = queuesa[:0]
+		groups  = map[string]struct{}{}
+		accName = c.acc.Name
+	)
+	if doQueues {
+		// Order the gws by lowest RTT
+		sort.Slice(gws, func(i, j int) bool {
+			return gws[i].getRTTValue() < gws[j].getRTTValue()
+		})
+	}
+	for _, gwc := range gws {
+		// hasInterest is not covering queue subs
+		psubInterest := gwc.hasInterest(accName, subj)
+		if !psubInterest && !doQueues {
```

Comment:
so if doQueues do not need to evaluate hasInterest, correct?

Reply:
You can send to a queue group locally but have other queue groups across a GW.

Reply:
I think we have a problem indeed.

Reply:
No problem, write the test first - lol. Seriously though I think we gather the queue groups as a suppression list from local (incl routes) deliveries and pass to the function to deliver to GWs.

Reply:
I will, but it is clear that as the code stands now, the server would not try to deliver to group foo.baz to remote GW just because it had sent to foo.bar locally.

Reply:
Yes I think we need to fix that. Do the test first though ;)